airbyte-integrations/connectors/source-shopify/source_shopify/streams/base_streams.py

```diff
@@ -110,11 +111,15 @@ def produce_records(
 
     def should_retry(self, response: requests.Response) -> bool:
         known_errors = ShopifyNonRetryableErrors(self.name)
+        connection_errors = ShopifyRetryableErrors(self.name)
         status = response.status_code
         if status in known_errors.keys():
             setattr(self, "raise_on_http_errors", False)
             self.logger.warning(known_errors.get(status))
             return False
+        elif status in connection_errors:
```

Comment:
I still think that this code path will never be executed in production as there are no responses and hence `should_retry` will not be called. Here is the trace that makes me think that:

```
>>> response = requests.get("http://localhost:5500")
Traceback (most recent call last):
  File "/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-salesforce/.venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 793, in urlopen
    response = self._make_request(
  File "/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-salesforce/.venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 537, in _make_request
    response = conn.getresponse()
  File "/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-salesforce/.venv/lib/python3.9/site-packages/urllib3/connection.py", line 466, in getresponse
    httplib_response = super().getresponse()
  File "/Users/maxime/.pyenv/versions/3.9.11/lib/python3.9/http/client.py", line 1377, in getresponse
    response.begin()
  File "/Users/maxime/.pyenv/versions/3.9.11/lib/python3.9/http/client.py", line 320, in begin
    version, status, reason = self._read_status()
  File "/Users/maxime/.pyenv/versions/3.9.11/lib/python3.9/http/client.py", line 281, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/Users/maxime/.pyenv/versions/3.9.11/lib/python3.9/socket.py", line 704, in readinto
    return self._sock.recv_into(b)
ConnectionResetError: [Errno 54] Connection reset by peer
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-salesforce/.venv/lib/python3.9/site-packages/requests/adapters.py", line 486, in send
    resp = conn.urlopen(
  File "/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-salesforce/.venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 847, in urlopen
    retries = retries.increment(
  File "/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-salesforce/.venv/lib/python3.9/site-packages/urllib3/util/retry.py", line 470, in increment
    raise reraise(type(error), error, _stacktrace)
  File "/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-salesforce/.venv/lib/python3.9/site-packages/urllib3/util/util.py", line 38, in reraise
    raise value.with_traceback(tb)
  File "/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-salesforce/.venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 793, in urlopen
    response = self._make_request(
  File "/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-salesforce/.venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 537, in _make_request
    response = conn.getresponse()
  File "/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-salesforce/.venv/lib/python3.9/site-packages/urllib3/connection.py", line 466, in getresponse
    httplib_response = super().getresponse()
  File "/Users/maxime/.pyenv/versions/3.9.11/lib/python3.9/http/client.py", line 1377, in getresponse
    response.begin()
  File "/Users/maxime/.pyenv/versions/3.9.11/lib/python3.9/http/client.py", line 320, in begin
    version, status, reason = self._read_status()
  File "/Users/maxime/.pyenv/versions/3.9.11/lib/python3.9/http/client.py", line 281, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/Users/maxime/.pyenv/versions/3.9.11/lib/python3.9/socket.py", line 704, in readinto
    return self._sock.recv_into(b)
urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(54, 'Connection reset by peer'))
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/maxime/.pyenv/versions/3.9.11/lib/python3.9/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
  File "/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-salesforce/.venv/lib/python3.9/site-packages/requests/api.py", line 73, in get
    return request("get", url, params=params, **kwargs)
  File "/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-salesforce/.venv/lib/python3.9/site-packages/requests/api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
  File "/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-salesforce/.venv/lib/python3.9/site-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
  File "/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-salesforce/.venv/lib/python3.9/site-packages/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
  File "/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-salesforce/.venv/lib/python3.9/site-packages/requests/adapters.py", line 501, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionResetError(54, 'Connection reset by peer'))

>>> response
Traceback (most recent call last):
  File "/Users/maxime/.pyenv/versions/3.9.11/lib/python3.9/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
NameError: name 'response' is not defined
```

What makes you think this code will be executed when there is a connection issue?

Reply:
To be honest, I'm not sure if this adds value. @bazarnov ? Should we drop this PR then?

Reply:
It's super weird because for the `source-mailchimp`, this `104` is retrievable by default somehow.
Check this: 
```
Backing off _send(...) for 5.0s (requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer')))
2024-04-25 16:54:23 source > Caught retryable error '('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))' after 1 tries. Waiting 5 seconds then retrying...
```
For: 
https://cloud.airbyte.com/workspaces/cb20618a-7501-4b4f-bc28-90ccd734007d/connections/1053c865-f988-420b-b075-260eeb266873/job-history

I've checked the code for declarative source and it doesn't seem to be handled elsewhere in the CDK not in the source.

Reply:
The issue is that we're not using any for that [here](https://github.com/airbytehq/airbyte/blob/6ea66e5770c8643a5b0a44b6bcab3806635cb13b/airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py#L326-L327) and this is where the failures come from ([example](https://cloud.airbyte.com/workspaces/6bc108e0-30fc-48aa-85e6-d967176a7f4e/connections/f8caa224-97d5-4661-a229-823572f98aa0/job-history#10050914::0)). `should_retry` is only called for HTTP requests within the HttpStream. Hopefully the new HttpRequestSender that Patrick is working on should remove some of the confusion as long as we use it everywhere

Reply:
So what do we do? If this change does no harm, but may have a potential benefit I'd say we merge it. But I'm OK merging/not merging.

Reply:
I agree this is probably "prod safe" but I think there is harm in thinking that errors are handled by this code and that a dev can just add such error handling here. If we decide the merge as is, I would like:
* A plan to track that this code path is applied in prod (can be a log with a weekly reminder to check
* An explicit comment that we think this code might not do anything and a description of the plan to confirm that with link so that we can keep track

Reply:
Let's go with this one:
> An explicit comment that we think this code might not do anything and a description of the plan to confirm that with link so that we can keep track

Because this one:
> A plan to track that this code path is applied in prod (can be a log with a weekly reminder to check

Will require CDK updates as well.

Reply:
Cool! For the second part, I don't think we need a CDK change. We change just have a log and check the Google Logs Explorer periodically. If once we have a connection aborted, we can check if the logs is there